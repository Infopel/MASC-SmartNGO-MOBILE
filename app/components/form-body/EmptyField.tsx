import { View } from "native-base"
import * as React from "react"

export interface FormControlFieldMapperProps {
}

/**
 * Describe your component here
 */
export const EmptyField = function (props: FormControlFieldMapperProps) {
    return (
        <View accessibilityLabel="Empty field" />
    )
}
