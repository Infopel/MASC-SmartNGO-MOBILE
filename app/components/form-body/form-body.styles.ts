import { StyleSheet } from "react-native"
import { color, typography,spacing,timing } from "../../theme"

export const FormBodyStyles = StyleSheet.create({
  container: { padding: spacing.medium },
  text: { fontFamily: typography.primary, fontSize: 14, color: color.primary },
})
