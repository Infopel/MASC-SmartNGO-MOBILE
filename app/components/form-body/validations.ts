import {
  IField,
  IBaseSubFormField,
  ITextField,
  IImageCaptureField,
} from 'store/form';
import * as yup from 'yup';

export function generateValidationSchema(fields: IField[]): yup.AnySchema {
  return yup.object().shape({
    ...fields.reduce((acc, v) => {
      let validation: Record<string, yup.AnySchema> = {};
      switch (v.type) {
        case 'text':
          validation[v.id] = textValidations(v);
          break;
        case 'sub-form':
          validation[v.id] = subformValidations(v);
          break;

        case 'image':
          validation[v.id] = imageValidations(v);
          break;
        default:
          validation[v.id] = baseValidations(v);
          break;
      }

      if (v.validation.required) validation[v.id] = validation[v.id].required();
      return {...acc, ...validation};
    }, {}),
  });
}

function baseValidations(field: IField): yup.BooleanSchema | yup.StringSchema {
  const val = yup.string();
  return val;
}
function textValidations(
  field: ITextField,
): yup.StringSchema | yup.NumberSchema {
  const {
    validation: {contentType, max, min},
  } = field;

  return contentType === 'number-positive' || contentType === 'decimal-positive'
    ? numberContentValidations(field)
    : textContentValidations(field);
}
function textContentValidations(
  field: ITextField,
): yup.StringSchema | yup.NumberSchema {
  const {
    validation: {contentType, max, min},
  } = field;
  let val: yup.StringSchema;

  val = yup.string();

  if (max) {
    val = val.max(max);
  }
  if (min) {
    val = val.min(min);
  }

  return val;
}
function numberContentValidations(
  field: ITextField,
): yup.StringSchema | yup.NumberSchema {
  const {
    validation: {contentType, max, min},
  } = field;
  let val: yup.NumberSchema = yup.number().typeError('Deve ser um numero');
  if (contentType === 'number-positive') {
    val = val.positive('Deve ser numero inteiro positivo');
  } else {
    val = val.positive('Deve ser numero positivo');
  }

  val = val.lessThan(max ?? Number.MAX_SAFE_INTEGER);

  val = val.moreThan(min ?? Number.MIN_SAFE_INTEGER);

  return val;
}

function subformValidations(field: IBaseSubFormField): yup.AnySchema {
  return yup.array().of(generateValidationSchema(field.fields));
}
function imageValidations(field: IImageCaptureField): yup.AnySchema {
  return yup.array().of(yup.string());
}
