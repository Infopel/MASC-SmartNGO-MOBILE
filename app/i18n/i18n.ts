import * as Localization from "expo-localization"
//@ts-ignore
import {I18n} from "i18n-js/dist/require"
import en from "./en"
import pt from "./pt"
export const i18n =  new I18n({en,pt})
i18n.enableFallback = true
i18n.defaultLocale = "en"
// i18n.missingBehavior = "error";

i18n.locale = Localization.locale || "en"

/**
 * Builds up valid keypaths for translations.
 * Update to your default locale of choice if not English.
 */
type DefaultLocale = typeof en
export type TxKeyPath = RecursiveKeyOf<DefaultLocale>

type RecursiveKeyOf<TObj extends Record<string, any>> = {
  [TKey in keyof TObj & string]: TObj[TKey] extends Record<string, any>
  //@ts-ignore  
  ? `${TKey}` | `${TKey}.${RecursiveKeyOf<TObj[TKey]>}`
    : `${TKey}`
}[keyof TObj & string]
