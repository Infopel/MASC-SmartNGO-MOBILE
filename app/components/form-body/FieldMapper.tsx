import { FormControlDate } from "components/form-control-date/form-control-date"
import { FormControlImageCapture } from "components/form-control-image-capture/form-control-image-capture"
import { FormControlInput } from "components/form-control-input/form-control-input"
import { FormControlRadio } from "components/form-control-radio/form-control-radio"
import { FormControlSubForm } from "components/form-control-sub-form/form-control-sub-form"
import { FormControlSeletorContainer } from "containers/form-control-seletor/form-control-seletor"
import * as React from "react"
import { StyleProp, ViewStyle } from "react-native"
import { IField } from "store/form"
import { EmptyField } from "./EmptyField"


export interface FormControlFieldMapperProps {
    /**
     * An optional style override useful for padding & margin.
     */
    style?: StyleProp<ViewStyle>
    field: IField
    ignoreSubform?: boolean
    setScrollTo?: (y: number) => void
}

/**
 * Describe your component here
 */
export const FieldMapper = function ({ ignoreSubform = false, field, setScrollTo, ...rest }: FormControlFieldMapperProps) {

    const fieldProps = { fieldName: field.id, label: field.name, helperText: field.helperText, validation: field.validation }

    switch (field.type) {
        case 'text':
            return <FormControlInput testID={field.id} {...fieldProps} placeholder={field.placeholder} />
        case 'date':
            return <FormControlDate testID={field.id} {...fieldProps} placeholder={field.placeholder} />
        case 'selector':
            return <FormControlSeletorContainer fromQueryId={field.queryFrom} testID={field.id} {...fieldProps}  items={field.items} placeholder={field.placeholder} />
        case 'radio':
            return <FormControlRadio testID={field.id} {...fieldProps} items={field.items} />
        case 'image':
            return <FormControlImageCapture testID={field.id} {...fieldProps} />
        case 'sub-form':
            if (!ignoreSubform && setScrollTo)
                return <FormControlSubForm setScrollTo={setScrollTo} testID={field.id} {...fieldProps} />
        default:
            return <EmptyField />
    }
}
