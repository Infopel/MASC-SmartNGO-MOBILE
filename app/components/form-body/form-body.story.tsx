import * as React from "react"
import { storiesOf } from "@storybook/react-native"
import { StoryScreen, Story, UseCase } from "../../../storybook/views"
import { color } from "../../theme"
import { FormBody } from "./form-body.component"

storiesOf("FormBody", module)
  .addDecorator((fn) => <StoryScreen>{fn()}</StoryScreen>)
  .add("Style Presets", () => (
    <Story>
      <UseCase text="Primary" usage="The primary.">
        <FormBody style={{ backgroundColor: color.error }} />
      </UseCase>
    </Story>
  ))
