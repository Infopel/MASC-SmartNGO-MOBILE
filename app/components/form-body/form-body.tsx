import { Button } from "components/button/button"
import { Formik, FormikHelpers } from "formik"
import { TxKeyPath } from "i18n/i18n"
import { translate } from "i18n/translate"
import { Box, Button as NBbtn, FlatList, IStackProps, Text, View, VStack } from "native-base"
import { goBack } from "navigators/navigation-utilities"
import * as React from "react"
import { FlatList as FL, NativeScrollEvent, NativeSyntheticEvent } from "react-native"
import { IField } from "store/form"
import { FieldMapper } from "./FieldMapper"
import { generateValidationSchema } from "./validations"


export interface FormBodyProps extends IStackProps {
  isLastPage?: boolean
  isFirstPage?: boolean
  isMultiPage: boolean
  prevPage?: () => void
  submitForm: (data: any) => TxKeyPath | undefined
  nextPage?: () => void
  fields: IField[]
}

/**
 * Describe your component here
 */
export const FormBody = function (props: FormBodyProps) {
  const { style, fields, isFirstPage, submitForm, isMultiPage,isLastPage = true, prevPage, nextPage, ...rest } = props
  const [error, setError] = React.useState<TxKeyPath>()
  const [scrollPosition, setScrollPosition] = React.useState<number>(0)
  const listRef = React.useRef<FL>(null)
  const initialValues = fields.reduce((acc, { id }) => { acc[id] = undefined; return acc }, {} as any)
  const handleOnSubmit = async (values: any, helper: FormikHelpers<any>) => {
    setError(undefined)
    console.log({ values })
    if(isLastPage){
      const response = await submitForm(values)
      console.log({ response })
      if (!response) {
  
        goBack()
      }
      else {
  
        setError(response)
  
      }
    }else {
      nextPage?.()
    }
  }
  const handleOnScroll = (event: NativeSyntheticEvent<NativeScrollEvent>) => {
     setScrollPosition(event.nativeEvent.contentOffset.y)
  }

  function scrollToPresentSubform(y: number) {
    listRef.current?.scrollToOffset({
      animated: true, offset: scrollPosition+(y),
    })
  }
  

  return (
    <VStack {...rest} flex='1'>
      <Formik initialValues={initialValues} validationSchema={generateValidationSchema(fields)} onSubmit={handleOnSubmit}>
        {({ handleSubmit }) =>
          <VStack space='2' flex='1'>
            {error && <Box bg='contrastThreshold' p='2' borderRadius={4}><Text color='danger.500'>{translate(error)}</Text></Box>}
            <FlatList onScroll={handleOnScroll} ref={listRef} flex='1' ItemSeparatorComponent={() => <View mb='2' />} keyExtractor={(item) => item.id} data={fields} renderItem={({ item }) => <FieldMapper setScrollTo={scrollToPresentSubform} field={item} />} />
            <NBbtn.Group >{isMultiPage ? <Button disabled={isFirstPage} flex='1' tx='form.prev-page' onPress={prevPage} /> : <View />}<Button onPress={handleSubmit} flex='1' tx={isLastPage ? 'form.submit' : 'form.next-page'} /></NBbtn.Group>
          </VStack>
        }
      </Formik>
    </VStack>
  )
}
