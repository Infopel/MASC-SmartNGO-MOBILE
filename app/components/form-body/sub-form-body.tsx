import { Button } from "components/button/button"
import { FlatList, IStackProps, View, VStack } from "native-base"
import * as React from "react"
import { IField } from "store/form"
import { FieldMapper } from "./FieldMapper"


export interface SubformBodyProps extends IStackProps {
  submitForm: () => void
  fields: IField[]
  index: number
  fieldName: string
}

/**
 * Describe your component here
 */
export function SubformBody (props: SubformBodyProps) {
  const { style,fieldName, fields, submitForm, index, ...rest } = props

  const handleSubmit = async () => {
    submitForm()
  }
  return (
    <VStack {...rest} flex='1'>
      <VStack space='2' flex='1'>
        <FlatList data={fields.map((x) => ({ ...x, id: `${fieldName}[${index}]${x.id}` }))}
          flex='1'
          ItemSeparatorComponent={() => <View mb='2' />} keyExtractor={(item) => item.id} renderItem={({ item }) => <FieldMapper ignoreSubform field={item} />} />
        <Button onPress={handleSubmit} flex='1' tx={'form.submit'} />
      </VStack>

    </VStack>
  )
}
