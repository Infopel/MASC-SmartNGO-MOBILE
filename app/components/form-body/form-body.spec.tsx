import { translate } from "i18n/translate"
import * as React from "react"
import { IField } from "store/form"
import { fireEvent, render } from "utils/test-utils/test-utils"

import { FormBody } from "./form-body"



describe("<FormBody>", () => {
  it("render", async function () {
    const component = <FormBody fields={[]}></FormBody>
    expect(render(component).toJSON()).toMatchSnapshot()
  })

  it('should render finish button by default', function () {
    const { queryByText } = render(<FormBody fields={[]}></FormBody>)
    expect(queryByText(translate('form.submit'))).toBeTruthy()

  })

  it("should handle next page button if is not last page", function () {
    const buttonPress = jest.fn()
    const { getByText, queryByText } = render(<FormBody nextPage={buttonPress} isLastPage={false} fields={[]}></FormBody>)
    expect(queryByText(translate('form.submit'))).toBeFalsy()
    const button = getByText(translate('form.next-page'))
    fireEvent.press(button)
    expect(buttonPress).toBeCalled()
  })

  it("should not handle next page button if is last page", function () {
    const buttonPress = jest.fn()
    const { getByText, queryByText } = render(<FormBody nextPage={buttonPress} isLastPage={true} fields={[]}></FormBody>)
    expect(queryByText(translate('form.next-page'))).toBeFalsy()
    const button = getByText(translate('form.submit'))
    fireEvent.press(button)
    expect(buttonPress).not.toBeCalled()
  })

  it("should set previous button as disabled if is first page", () => {
    const buttonPress = jest.fn()
    const { getByText } = render(<FormBody prevPage={buttonPress} isFirstPage={true} isLastPage={false} fields={[]}></FormBody>)
    expect(getByText(translate('form.prev-page'))).toBeDisabled()

  })

  it("should not render previous page button", () => {
    const { getByText, queryByText } = render(<FormBody isLastPage={false} fields={[]}></FormBody>)
    expect(queryByText(translate('form.prev-page'))).toBeFalsy()

  })

  it("should handle previous page", function () {
    const buttonPress = jest.fn()
    const { getByText } = render(<FormBody prevPage={buttonPress} fields={[]}></FormBody>)
    const button = getByText(translate('form.prev-page'))
    fireEvent.press(button)
    expect(buttonPress).toBeCalled()
  })

  it("should render a text field", () => {
    const field: IField = { id: "text", name: "civic-incubator-details.title", type: "text" }
    const { getByText } = render(<FormBody fields={[field]} />)
    expect(getByText(translate(field.name))).toBeTruthy()
  })
  it("should render a date field", () => {
    const field: IField = { id: "text", name: "civic-incubator-details.title", type: "date" }
    const { getByText, debug } = render(<FormBody fields={[field]} />)
    expect(getByText(translate(field.name))).toBeTruthy()

  })
  it("should render a radiogroup field", () => {
    const field: IField = { id: "text", name: "civic-incubator-details.title", type: "radio", items:[{id:"sumba ud", name:"Non existent"},{id:"sba ud", name:"Existent"}] }
    const { getByText, debug } = render(<FormBody fields={[field]} />)
    expect(getByText(translate(field.name))).toBeTruthy()

  })
  it("should render a image capture field", () => {
    const field: IField = { id: "text", name: "civic-incubator-details.title", type: "image", items:[{id:"sumba ud", name:"Non existent"},{id:"sba ud", name:"Existent"}] }
    const { getByText, debug } = render(<FormBody fields={[field]} />)
    expect(getByText(translate(field.name))).toBeTruthy()

  })
  it("should render a selector field", () => {
    const field: IField = {
      id: "text", name: "civic-incubator-details.title", type: "selector", items: [{ id: "1", name: "Water" }, { id: "2", name: "Tomato" }]
    }
    const { getByText, getAllByText } = render(<FormBody fields={[field]} />)
    expect(getAllByText(translate(field.name)).length).toBe(2)
    expect(getByText((field.items[0].name))).toBeTruthy()
    expect(getByText((field.items[1].name))).toBeTruthy()
  })
  it("should render a list with all types output", () => {
    const fields: IField[] = [{
      id: "texssst", name: "civic-incubators.search", type: "text",
    },
    { id: "texsadt", name: "civic-incubator-details.title", type: "date", },
    {
      id: "textas", name: "new-civic-incubator.admin-post", type: "selector", items: [{ id: "1", name: "Water" }, { id: "2", name: "Tomato" }]
    }]

    const { getByText, getAllByText } = render(<FormBody fields={fields} />)
    expect(getByText(translate(fields[0].name))).toBeTruthy()
    expect(getByText(translate(fields[1].name))).toBeTruthy()
    expect(getAllByText(translate(fields[2].name)).length).toBe(2)

  })
})
