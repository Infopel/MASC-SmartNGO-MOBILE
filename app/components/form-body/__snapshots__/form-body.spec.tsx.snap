// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormBody> render 1`] = `
<View
  dataSet={Object {}}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    dataSet={Object {}}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      ItemSeparatorComponent={[Function]}
      contentContainerStyle={
        Array [
          Object {},
          Object {
            "dataSet": Object {},
          },
          Object {},
        ]
      }
      data={Array []}
      dataSet={Object {}}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "flex": 1,
        }
      }
      viewabilityConfigCallbackPairs={Array []}
    >
      <View />
    </RCTScrollView>
    <View
      dataSet={Object {}}
      style={
        Object {
          "height": 8,
        }
      }
    />
    <View
      dataSet={Object {}}
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        dataSet={Object {}}
        style={Object {}}
      />
      <View
        dataSet={Object {}}
        style={
          Object {
            "width": 8,
          }
        }
      />
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        colorScheme="primary"
        dataSet={Object {}}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#d37b16",
            "borderRadius": 4,
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "paddingBottom": 10,
            "paddingLeft": 12,
            "paddingRight": 12,
            "paddingTop": 10,
          }
        }
      >
        <View
          dataSet={Object {}}
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
          test={true}
        >
          <View
            dataSet={Object {}}
            style={Object {}}
          >
            <Text
              dataSet={Object {}}
              style={
                Object {
                  "backgroundColor": undefined,
                  "color": "#fafafa",
                  "fontFamily": undefined,
                  "fontSize": 14,
                  "fontStyle": "normal",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                  "lineHeight": 21,
                  "textDecorationLine": undefined,
                }
              }
            >
              Finish
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
